### PR TITLE
Image block: Revise lightbox UI to remove 'behaviors'

### LINKED ISSUE #53403: Image block: Revise Lightbox UI to remove reference to Behaviors
## What problem does this address?

As we've explored implementing an [image lightbox](https://github.com/WordPress/gutenberg/issues/51132), it's become increasingly clear that the feature is specific to the image, so we should revise the UI to be tailored to the image block accordingly instead of being encompassed under the [concept of Behaviors](https://github.com/WordPress/gutenberg/issues/50029).

## What is your proposed solution?

Let's design a simplified UX for the lightbox that makes sense in the context of just the image block.

Questions:
1. How are we going to surface the lightbox as a setting of the image block?
2. How are users going to activate the lightbox?
3. Will the lightbox come activated by default?
4. Will it be a global setting for all images?
5. How will the user deactivate the lightbox on a specific Image?

### PR DESCRIPTION
**Pull request by artemiomorales**
## What?
Now that the concept of [behaviors](https://github.com/WordPress/gutenberg/issues/50029) is no longer shipping in the near term, this PR removes it from the `theme.json`, the Global Styles and the Image block attributes.

## Why?
We want to provide a lightbox feature and get feedback on its implementation, but we also want to do so in a way that's minimal and easy to understand. Addresses #53403 

## How?
Removing all the references to `behaviors`. 

_Note regarding backwards compatibility: We've added backwards compatibility [in the following PR](https://github.com/WordPress/gutenberg/pull/54071), which is meant to be merged at the same time as this one so there's no gap in support for users using the old syntax we're removing in this PR._

## Testing Instructions
1. Open a post or page.
2. Insert an Image block.
3. Make sure that no "Behaviors" panel is preset inside of the "Advanced" panel of the Inspector Controls.
4. Save the post/page
5. Go to the frontend
6. Make sure that the page loads correctly and there is no Lightbox present.

### CODE REVIEW

File: packages/block-editor/src/components/global-styles/hooks.js
Comment (michalczaplinski) [inline packages/block-editor/src/components/global-styles/hooks.js:572]: If I'm not mistaken, we'll always want to show the "Behaviors" Panel (under a different name probably) in the UI now that the `theme.json` support is disabled. So this hook is not necessary anymore.

File: lib/block-supports/behaviors.php
Comment (michalczaplinski) [inline lib/block-supports/behaviors.php]: Should we not use the `::get_user_data()` method here after all?

Comment (artemiomorales) [reply lib/block-supports/behaviors.php]: Whoops, I did this wrong. Thanks for the catch!

**Comment by michalczaplinski** [comment]
I've pushed some last changes. There are a couple of things that are not working and we need final decision on how to structure the options. I've explained the changes here:

<div>
    <a href="https://www.loom.com/share/7c970f1314714a55b1ee46ff5afaa3bc">
      
    </a>
  </div>

File: packages/block-library/src/image/deprecated.js
Comment (michalczaplinski) [inline packages/block-library/src/image/deprecated.js:637]: This one I'm not 100% sure about. 

I think that we can safely remove the reference to behaviors in those deprecations because if a user does have a block with the `behaviors: { ...}` attribute for some reason, it should just be ignored. 

Could you confirm that @gziolo ?

File: packages/block-library/src/image/index.php
Comment (michalczaplinski) [inline packages/block-library/src/image/index.php:39]: I'm assuming that this will be the structure of the "new" attribute that controls the lightbox.

File: packages/block-library/src/image/index.php
Comment (michalczaplinski) [inline packages/block-library/src/image/index.php:266]: Noting the TODO for the subsequent PR that adds the revised Lightbox implementation.

File: packages/block-library/src/image/index.php
Comment (michalczaplinski) [inline packages/block-library/src/image/index.php:86]: As in the previous comment, I'm also assuming that this will be the structure of the "new" attribute that controls the lightbox.

File: test/e2e/specs/editor/blocks/image.spec.js
Comment (michalczaplinski) [inline test/e2e/specs/editor/blocks/image.spec.js:738]: I'm going to skip this test for now because we'll have to adjust the tests in the subsequent PR to use the "new" Lightbox UI.

File: test/e2e/specs/editor/blocks/image.spec.js
Comment (michalczaplinski) [inline test/e2e/specs/editor/blocks/image.spec.js:933]: There's no fade animation anymore (only "expand") so we can delete this test.

File: test/e2e/specs/editor/blocks/image.spec.js
Comment (michalczaplinski) [inline test/e2e/specs/editor/blocks/image.spec.js]: Removing this test as well, because there is no "animation selector" anymore. The Lightbox only has the "zoom" effect (animation) - it's the only one.

File: packages/block-library/src/image/deprecated.js
Comment (gziolo) [reply packages/block-library/src/image/deprecated.js:637]: Block supports can be safely removed. The only question is what will happen when there is a value set for the `behaviors` block attribute set in the saved HTML as part of the comment metadata. As far as I can tell, it was: `behaviors.lightbox` (based on `$attributes['behaviors']['lightbox']`) so it would be nice to transfer the value to whatever else you plan to use moving forward. 

One way to test whether renaming the attribute is safe, would be pasting the old saved Image block with the lightbox enabled locally and checking whether the block editor won't trigger the block validation error. You can also simulate that by adding a new block fixture in integration tests as explained in https://github.com/WordPress/gutenberg/blob/trunk/test/integration/fixtures/blocks/README.md.

File: lib/theme.json
Comment (michalczaplinski) [inline lib/theme.json:12]: Can this just be removed or does it need a deprecation of some sort? Behaviors were [shipped](https://github.com/WordPress/gutenberg/releases/tag/v16.4.0) in GB 16.4 so never included in a WP release.

cc @ellatrix @gziolo @youknowriad

Comment (gziolo) [reply lib/theme.json:12]: @artemiomorales and I looked at whether this is used in production, and it looks like you can change that setting in the UI, which most likely gets reflected in what gets saved in the database. While it isn't mandatory for WordPress core, it would still be nice to provide a migration path for sites that have the latest Gutenberg plugin enabled. It shouldn't be that difficult to code something in the plugin to read `behaviors` object from the site and transfer the value to the new structure.

I guess the same applies to the block attribute that changes from `behaviors.lightbox` to `lightbox` in the Image block.

Comment (oandregal) [reply lib/theme.json:12]: I'm not fully caught up on what's next for behaviors, but it seems that themes will no longer be able to tweak the lightbox feature (enable/disable, set animation) after this removal.

If we still plan to allow them to tweak it, just in a different place, we could add some code around [this line](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L617) to make the migration easier for them. I'll be AFK in the next few days, but wanted to share in case it's useful.

Comment (michalczaplinski) [reply lib/theme.json:12]: In the followup PR: https://github.com/WordPress/gutenberg/pull/54071 we've added backwards-compatibility for users that might have been using the previous `behaviors.lightbox` syntax. 

In that PR (which is based on the current one) we're also now introducing a `lightbox.allowEditing` setting which allows themes to enable/disable the lightbox.

**Comment by ramonjd** [comment]
I've taken this PR for a spin, and checked the global styles/revisions rest API changes and associated tests. From that perspective, it looks good to me. 

On the backwards compatibility subject - I've heard it more than once that the plugin itself supports the previous two versions only, when those changes do not impact on Core. 

Here's that noted in a comment: https://github.com/WordPress/gutenberg/pull/47475#issuecomment-1409560457

But I can't yet find anything official. I think it's because there is nothing official. 

Was this published as an experimental API? The use of `__experimental` suggest yes. Will it affect anyone who updates to the latest WordPress version? Probably not, right? These are some things we could ask ourselves in two versions' time.

I guess we should be always wary when it's a public API. It was a topic discussed at the WordPress community summit this year. See: https://make.wordpress.org/summit/2023/08/23/community-summit-discussion-notes-addressing-backwards-compatibility-in-gutenberg/

Related issue:

- https://github.com/WordPress/gutenberg/issues/41265

File: test/e2e/specs/editor/blocks/image.spec.js
Comment (ramonjd) [reply test/e2e/specs/editor/blocks/image.spec.js:738]: Would be it okay to add a comment here for people like me who forget stuff (like looking at Github annotations) 😄 ?

**Comment by michalczaplinski** [comment]
Thanks a lot for the review, Ramon! 

> Was this published as an experimental API? The use of __experimental suggest yes.
The API itself can be considered experimental, but since it's exposed in the UI, there's not really a way for users to know if what they are enabling is experimental or not. It was not explicitly described as "experimental" in the UI. 

It was never shipped in Core, though, so I guess deprecation is the way to go!

Just to let you know, there is a follow-up PR to this one: https://github.com/WordPress/gutenberg/pull/54071/
There, we add the updated version of the Lightbox (without behaviors).

File: packages/block-library/src/image/deprecated.js
Comment (ajlende) [reply packages/block-library/src/image/deprecated.js:637]: We should probably add a block deprecation for this either way, right? If there is no validation error, you should be able to set an `isElegible` method on the deprecation based on the presence of the top-level `behaviors` attribute. And then the `behaviors` attribute can be migrated to the `lightbox` attribute with the `migrate` method.

This would save having to add the deprecation warning in https://github.com/WordPress/gutenberg/pull/54071/files#diff-a14b61c874095649b236b78413e805b4541277b24eb71ce350a17d9f2b6396b1R765-R779.

Comment (artemiomorales) [reply packages/block-library/src/image/deprecated.js:637]: > One way to test whether renaming the attribute is safe, would be pasting the old saved Image block with the lightbox enabled locally and checking whether the block editor won't trigger the block validation error.

I've tested this by adding the JSON manually to the block and there's no error, so it looks like renaming the attribute is safe.

> We should probably add a block deprecation for this either way, right?

@ajlende This sounds right.

As per [this comment](https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575), it looks like we can handle the migration in a separate PR later, right? 

We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

File: test/e2e/specs/editor/blocks/image.spec.js
Comment (artemiomorales) [reply test/e2e/specs/editor/blocks/image.spec.js:738]: Done 👍

**Review by ramonjd — APPROVED** [review]
I think the state of this PR is good, thanks for sticking with it. 

If @gziolo or @ajlende don't have any further concerns, then it LGTM.

I understand this needs to be merged before https://github.com/WordPress/gutenberg/pull/54071, so I suppose we could address follow up concerns in that PR?

**Comment by ajlende** [comment]
I tested with the following block content and theme.json modification.

```html
<!-- wp:heading -->
<h2 class="wp-block-heading">Global</h2>
<!-- /wp:heading -->

<!-- wp:image {"id":68,"sizeSlug":"large","linkDestination":"none"} -->
<figure class="wp-block-image size-large"><img src="http://wp.local/wp-content/uploads/2023/02/27982606245_0a702a9b26_o-1024x576.jpg" alt="" class="wp-image-68"/></figure>
<!-- /wp:image -->

<!-- wp:heading -->
<h2 class="wp-block-heading">Block</h2>
<!-- /wp:heading -->

<!-- wp:image {"id":68,"sizeSlug":"large","linkDestination":"none","behaviors":{"lightbox":{"enabled":true,"animation":"fade"}}} -->
<figure class="wp-block-image size-large"><img src="http://wp.local/wp-content/uploads/2023/02/27982606245_0a702a9b26_o-1024x576.jpg" alt="" class="wp-image-68"/></figure>
<!-- /wp:image -->
```

```jsonc
{
	// ...
	"behaviors": {
		"blocks": {
			"core/image": {
				"lightbox": {
					"enabled": true
				}
			}
		}
	},
	// ...
}
```

This is what I see.

| Before | After |
| ------ | ----- |
| behaviors-deprecation-lightbox-before | behaviors-deprecation-lightbox-after |
| behaviors-deprecation-lightbox-global-before | behaviors-deprecation-lightbox-global-after |
| behaviors-deprecation-lightbox-block-before | behaviors-deprecation-lightbox-block-after |

> As per [this comment](https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575), it looks like we can handle the migration in a separate PR later, right?

@artemiomorales The lightbox support wasn't hidden behind an experimental flag or anything in the UI, so folks using the plugin who have been using lightbox will see this as breaking their site if you don't have the migration. So I would recommend getting the migration handled before merging this PR so we can be _sure_ they make it into the same GB release.

> We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

As this PR is right now, there is no backwards compatibility. The functionality has simply been removed.

We have systems for migrating both block attributes and global styles settings that should be used.

For block attributes, you just need to migrate the attributes to their new location with a block deprecation. Since the block supports code is removed, you'll have to replace the supports in the block deprecations with the `behaviors` attribute in `attributes`. Then you can use the `isEligible` method to trigger the migration.

For global styles (theme.json), we also have a migration system in class-wp-theme-json-schema.php. You can write a migration function to move the setting where you wish.

The block supports part is the only part where you'll need to properly add a deprecation. That can be done simply by adding a [`_deprecated_function()`](https://developer.wordpress.org/reference/functions/_deprecated_function/) call to the functions and removing only the block support registration.

Any public function you want to remove can't be deleted if you're providing backwards compatibility. Instead, you add one of our deprecated functions that will print a warning. Then in two versions, the function can finally be removed.

File: packages/block-library/src/image/deprecated.js
Comment (ajlende) [reply packages/block-library/src/image/deprecated.js:637]: > As per https://github.com/WordPress/gutenberg/pull/54071#discussion_r1318511575, it looks like we can handle the migration in a separate PR later, right?
>
> We'll be offering backwards compatibility for 2 Gutenberg releases. As this PR is already pretty big, I think it makes sense to just push this forward and handle the deprecation separately.

[Responded below.](https://github.com/WordPress/gutenberg/pull/53851#issuecomment-1712082579)

**Comment by michalczaplinski** [comment]
The backwards compatibility concerns have been addressed in https://github.com/WordPress/gutenberg/pull/54509 👍

**Pull request merged**
